.github/actions/cluster/lxd-cluster.sh

```diff
@@ -0,0 +1,322 @@
+        if [ ! "${exists}" ]; then
+                lxc storage create "${STORAGE_POOL}" zfs
+        fi
+
+        # Setup cluster VMs.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+
+                state=$(lxc list --format csv --columns s "${instance}")
+                case "${state}" in
+                "RUNNING")
+                        echo "Instance ${instance} already running."
+                        continue
+                        ;;
+                "STOPPED")
+                        echo "Starting instance ${instance}..."
+                        lxc start "${instance}"
+                        continue
+                        ;;
+                esac
+
+                args=""
+                if [ "${INSTANCE_TYPE}" = "virtual-machine" ]; then
+                        args="--vm"
+                else
+                        args="-c security.nesting=true"
+                fi
+
+                echo "Creating instance ${instance} ..."
+
+                lxc launch "${INSTANCE_IMAGE}" "${instance}" \
+                        --storage "${STORAGE_POOL}" \
+                        --network "${NETWORK_NAME}" \
+                        -c limits.cpu=4 \
+                        -c limits.memory=4GiB \
+                        $args
+        done
+
+        # Wait for instances to become ready.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+                waitInstance "${instance}"
+                lxc exec "${instance}" -- systemctl is-system-running --wait
+        done
+
+        # Install LXD on VMs.
+        for i in $(seq 1 "${CLUSTER_SIZE}"); do
+                instance="${INSTANCE}-${i}"
+
+                echo "Preparing instance ${instance} ..."
+
+                # Install snap daemon.
+                lxc exec "${instance}" --env=DEBIAN_FRONTEND=noninteractive -- apt-get -qq -y install snapd
+
+                # Install LXD snap.
+                lxc exec "${instance}" -- snap install lxd --channel "${VERSION_LXD}" || lxc exec "${instance}" -- snap refresh lxd --channel "${VERSION_LXD}"
+
+                hasBucketSupport=$(lxc exec "${instance}" -- lxc info | grep -e "- storage_buckets" || true)
+
+                # Install MinIO if enabled.
+                if [ "${MINIO_ENABLED}" = "true" ] && [ "${hasBucketSupport}" != "" ]; then
+                        echo "Installing MinIO server and client on instance ${instance} ..."
+
+                        lxc exec "${instance}" -- mkdir -p "${MINIO_INSTALL_DIR}"
+
+                        # Download latest minio server and client.
+                        lxc exec "${instance}" -- curl -sSfL https://dl.min.io/server/minio/release/linux-amd64/minio --output "${MINIO_INSTALL_DIR}/minio"
+                        lxc exec "${instance}" -- curl -sSfL https://dl.min.io/client/mc/release/linux-amd64/mc --output "${MINIO_INSTALL_DIR}/mc"
```

Comment:
Please move those out of the loop so they are downloaded only once per cluster and then `lxc push`'ed into the instance.